File: src/main/java/io/chesslave/model/Piece.java
package io.chesslave.model;

import io.chesslave.Functions;
import javaslang.collection.List;
import javaslang.collection.Set;
import java.util.Objects;

/**
 * A chess piece.
 */
public class Piece {

    public enum Type {

        PAWN,
        KNIGHT,
        BISHOP,
        ROOK,
        QUEEN,
        KING
    }

    public final Type type;
    public final Color color;

    private Piece(Type type, Color color) {
        this.type = type;
        this.color = color;
    }

    public static Piece of(Type type, Color color) {
        return new Piece(type, color);
    }

    /**
     * @return All chess pieces available.
     */
    public static Set<Piece> all() {
        return List.of(Piece.Type.values()).crossProduct(List.of(Color.values()))
                .map(Functions.of(Piece::new).tupled())
                .toSet();
    }

    /**
     * @return True if the given piece has the same color of this piece.
     */
    public boolean isFriend(Piece piece) {
        return this.color == piece.color;
    }

    /**
     * @return True if the given piece has not the same color of this piece.
     */
    public boolean isOpponent(Piece piece) {
        return this.color != piece.color;
    }

    @Override
    public String toString() {
        return String.format("%s %s", type, color);
    }

    @Override
    public boolean equals(Object rhs) {
        if (rhs instanceof Piece == false) {
            return false;
        }
        final Piece other = (Piece) rhs;
        return this.type == other.type &&
                this.color == other.color;
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, color);
    }
}

File: backend/src/main/java/io/chesslave/model/Piece.kt
package io.chesslave.model

import javaslang.collection.HashSet
import javaslang.collection.Set

enum class Color {

    WHITE,
    BLACK;

    fun opponent(): Color = if (this == WHITE) BLACK else WHITE
}

/**
 * A chess piece.
 */
data class Piece(val type: Type, val color: Color) {

    enum class Type(val value: Int) {
        PAWN(1),
        KNIGHT(3),
        BISHOP(3),
        ROOK(5),
        QUEEN(9),
        KING(0)
    }

    /**
     * @return True if the given piece has the same color of this piece.
     */
    fun isFriend(piece: Piece): Boolean = this.color === piece.color

    /**
     * @return True if the given piece has not the same color of this piece.
     */
    fun isOpponent(piece: Piece): Boolean = !isFriend(piece)

    companion object {

        val whitePawn = pawnOf(Color.WHITE)
        val whiteBishop = bishopOf(Color.WHITE)
        val whiteKnight = knightOf(Color.WHITE)
        val whiteRook = rookOf(Color.WHITE)
        val whiteQueen = queenOf(Color.WHITE)
        val whiteKing = kingOf(Color.WHITE)
        val blackPawn = pawnOf(Color.BLACK)
        val blackBishop = bishopOf(Color.BLACK)
        val blackKnight = knightOf(Color.BLACK)
        val blackRook = rookOf(Color.BLACK)
        val blackQueen = queenOf(Color.BLACK)
        val blackKing = kingOf(Color.BLACK)

        fun pawnOf(color: Color) = Piece(Type.PAWN, color)
        fun bishopOf(color: Color) = Piece(Type.BISHOP, color)
        fun knightOf(color: Color) = Piece(Type.KNIGHT, color)
        fun rookOf(color: Color) = Piece(Type.ROOK, color)
        fun queenOf(color: Color) = Piece(Type.QUEEN, color)
        fun kingOf(color: Color) = Piece(Type.KING, color)

        /**
         * All chess pieces available.
         */
        val all: Set<Piece>
            = Type.values()
            .flatMap { type -> Color.values().map { color -> Piece(type, color) } }
            .let { HashSet.ofAll(it) }
    }
}